### PR TITLE
KEP-1880 Add ut for pkg/registry/networking/servicecidr

### PR DESCRIPTION
**Pull request by bzsuni**
#### What type of PR is this?
/kind cleanup
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs
```

### CODE REVIEW

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 1]
```diff
@@ -15,3 +15,117 @@ limitations under the License.
...
+func TestPrepareForUpdate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	newObj := &networking.ServiceCIDR{
+		Spec: networking.ServiceCIDRSpec{
+			CIDRs: []string{"10.10.0.0/24"},
+		},
+	}
+	oldObj := &networking.ServiceCIDR{}
+	strategy.PrepareForUpdate(context.TODO(), newObj, oldObj)
+	// Assert that cleared fields are now empty/unset
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:66]: what is this testing?

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 2]
```diff
@@ -15,3 +15,117 @@ limitations under the License.
...
+		fieldpath.MakePathOrDie("status"),
+	)
+	if !fields["networking/v1alpha1"].Equals(status) {
+		t.Errorf("Expected 'status' field to be reset for networking/v1alpha1")
+	}
+}
+
+func TestPrepareForCreate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	obj := &networking.ServiceCIDR{Status: networking.ServiceCIDRStatus{}}
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:50]: should this test that this has something in Status and is cleared?

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 3]
```diff
@@ -15,3 +15,117 @@ limitations under the License.
...
+func TestGetResetFields(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	fields := strategy.GetResetFields()
+	status := fieldpath.NewSet(
+		fieldpath.MakePathOrDie("status"),
+	)
+	if !fields["networking/v1alpha1"].Equals(status) {
+		t.Errorf("Expected 'status' field to be reset for networking/v1alpha1")
+	}
+}
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:46]: I prefer the existing mode in other places that does not depend on the api group and version

```go
func TestStrategy_ResetFields(t *testing.T) {
	resetFields := Strategy.GetResetFields()
	if len(resetFields) != 1 {
		t.Errorf("ResetFields should have 1 element, but have %d", len(resetFields))
	}
}
```

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 4]
```diff
@@ -15,3 +15,117 @@ limitations under the License.
...
+
+func TestValidateUpdate(t *testing.T) {
+	strategy := serviceCIDRStrategy{}
+	newObj := &networking.ServiceCIDR{Spec: networking.ServiceCIDRSpec{CIDRs: []string{"bad cidr"}}}
+	oldObj := &networking.ServiceCIDR{}
+	errors := strategy.ValidateUpdate(context.TODO(), newObj, oldObj)
+	if len(errors) == 0 {
+		t.Errorf("Expected validation errors for invalid update")
+	}
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:85]: is this not an error?

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 2 above)
Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:50]: Should some cleaning logic be added to the method being tested?
```go
// PrepareForCreate clears the status of an ServiceCIDR before creation.
func (serviceCIDRStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
	_ = obj.(*networking.ServiceCIDR)

}
```

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 3 above)
Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:46]: Yes , it's better.

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 4 above)
Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:85]: No?

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 2 above)
Comment (aojea) [reply pkg/registry/networking/servicecidr/strategy_test.go:50]: yep

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 4 above)
Comment (aojea) [reply pkg/registry/networking/servicecidr/strategy_test.go:85]: it should be, the newObj contains an invalud cidr  🤔

Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:85]: > it should be, the newObj contains an invalud cidr 🤔

The assertion fails if the length of error is zero in this test. This is as expected

Comment (aojea) [reply pkg/registry/networking/servicecidr/strategy_test.go:85]: oh, lol ... then assert on the exact number of errors

Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:49]: Yes

Comment (aojea) [reply pkg/registry/networking/servicecidr/strategy_test.go:49]: this is still missing, I think it should be something as

```
if len(errors) != 1 { 
  t.Errorf("Expected 1 validation errors for invalid update, got %d", len(errors))
```

Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:49]: Making precise assertions is not easy to achieve.
If the test code is
```go
oldObj := newServiceCIDR()
newObj := oldObj.DeepCopy()
newObj.Spec.CIDRs = []string{"bad cidr"}
if len(errors) != 1 {
	t.Errorf("Expected 1 validation errors for invalid update, got %d, errors: %v", len(errors), errors)
}
```
it got a log as
```shell
=== RUN   TestServiceCIDRStrategy
    /root/go/src/k8s.io/kubernetes/pkg/registry/networking/servicecidr/strategy_test.go:62: Expected 1 validation errors for invalid update, got 2, errors: [spec.cidrs[0]: Invalid value: "bad cidr": netip.ParsePrefix("bad cidr"): no '/' spec.cidrs: Invalid value: []string{"bad cidr"}: field is immutable]
--- FAIL: TestServiceCIDRStrategy (0.00s)
FAIL
FAIL    k8s.io/kubernetes/pkg/registry/networking/servicecidr   0.055s
```

Comment (aojea) [reply pkg/registry/networking/servicecidr/strategy_test.go:49]: ok, then 2 errors are expected , you can create a table test here if you want or just use 

```
if len(errors) != 2 { 
  t.Errorf("Expected 2 validation errors for invalid update, got %d", len(errors))
```

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 5]
```diff
@@ -15,3 +15,84 @@ limitations under the License.
...
+	resetFields := Strategy.GetResetFields()
+	if len(resetFields) != 1 {
+		t.Errorf("ResetFields should have 1 element, but have %d", len(resetFields))
+	}
+	obj := &networking.ServiceCIDR{Spec: networking.ServiceCIDRSpec{CIDRs: []string{"bad cidr"}}}
+
+	errors := Strategy.Validate(context.TODO(), obj)
+	if len(errors) == 0 {
+		t.Errorf("Expected validation errors for invalid object")
+	}
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:55]: same comment as before, assert on the exact number of errors we are expecting

**Comment by aojea** [comment]
one final comment and we are good to go

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 4 above)
Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go:63]: Done

File: pkg/registry/networking/servicecidr/strategy_test.go (hunk 5 above)
Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go]: Done, thank you very much.

File: pkg/registry/networking/servicecidr/strategy_test.go [hunk 6]
```diff
@@ -15,3 +15,84 @@ limitations under the License.
...
+
+	resetFields := Strategy.GetResetFields()
+	if len(resetFields) != 1 {
+		t.Errorf("ResetFields should have 1 element, but have %d", len(resetFields))
+	}
+	obj := &networking.ServiceCIDR{Spec: networking.ServiceCIDRSpec{CIDRs: []string{"bad cidr"}}}
+
+	errors := Strategy.Validate(context.TODO(), obj)
+	if len(errors) != 2 {
+		t.Errorf("Expected 1 validation errors for invalid object, got %d", len(errors))
```
Comment (aojea) [inline pkg/registry/networking/servicecidr/strategy_test.go:54]: you have a mismatch, or either 1 or 2 :)

Comment (bzsuni) [reply pkg/registry/networking/servicecidr/strategy_test.go]: oh, I'm sorry, it's done now, thank you very much

**Comment by aojea** [comment]
/lgtm
/approve

Thanks

**Pull request merged**